### PR TITLE
[VL] Support posexplode function and code refactoring on GenerateExecTransformer

### PR DESCRIPTION
**Pull request by marin-ma**
This patch contains 2 major changes:

- Support `posexplode` for Velox backend.
- Separate `GenerateExecTransformer` for Velox/CH backends. For Velox backend, the RelNode creation requires pre/projection for different generators. For CH backend, the RelNode creation is kept clean.

### CODE REVIEW

**Comment by marin-ma** [comment]
@zhouyuan @zzcclp Please kindly review. Thanks!

**Comment by zhouyuan** [comment]
@marin-ma Thanks for helping on this, can you please do a rebase? There's a small conflict on metrics

File: backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    if (outer) {
+      return ValidationResult.notOk(s"Velox backend does not support outer")
+    }
+    generator match {
+      case _: JsonTuple =>
+        ValidationResult.notOk(s"Velox backend does not support this json_tuple")
+      case e: ExplodeBase =>
+        e.child match {
+          case _: CreateMap =>
+            // explode(MAP(col1, col2))
```
Comment (zhouyuan) [inline backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala:74]: is this also necessary to posexplode(MAP()) ?

File: backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+          generateRel,
+          (requiredOutput ++ flattenStruct).asJava,
+          context,
+          operatorId,
+          1 + requiredOutput.size // 1 stands for the inner struct field from array.
+        )
+      case PosExplode(posExplodeChild) =>
+        // Ordinal populated by Velox UnnestNode starts with 1.
+        // Need to substract 1 to align with Spark's output.
+        val unnestedSize = posExplodeChild.dataType match {
```
Comment (zhouyuan) [inline backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala:202]: it looks MAP is supported here

**Review by zhouyuan — COMMENTED** [review]
looks good to me, just one question needs to check

File: backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala (hunk 1 above)
Comment (marin-ma) [reply backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala]: Looks like `CreateMap` is not covered by existing UTs. However, I verified locally and found Gluten does support this case. Removed this constraint in the latest commit and add UT https://github.com/apache/incubator-gluten/pull/4901/commits/131a08cdd8ade0edd0e7a58c2b401e213adb640e
cc: @liujiayi771

File: backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        .build()
+      val optimization = BackendsApiManager.getTransformerApiInstance.packPBMessage(message)
+      ExtensionBuilder.makeAdvancedExtension(optimization, null)
+    } else {
+      getExtensionNodeForValidation
+    }
+  }
+
+  // Select child outputs and append generator input.
+  private def applyPreProject(
```
Comment (liujiayi771) [inline backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala:146]: Could you refactor this part using `PullOutPreProject` rule?

File: backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      inputRel,
+      projectExpressions.asJava,
+      context,
+      operatorId,
+      child.output.size)
+  }
+
+  // There are 3 types of CollectionGenerator in spark: Explode, PosExplode and Inline.
+  // Adds postProject for PosExplode and Inline.
+  private def applyPostProject(
```
Comment (liujiayi771) [inline backends-velox/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala:170]: Ditto.

**Review by zhouyuan — APPROVED** [review]
👍 
Note, there will be a follow up to use to project pull out logic.

**Pull request merged**